### PR TITLE
methods-method_overloading-hard-q4

### PR DESCRIPTION
**Pull request by ketphan02**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [x] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/methods/method_overloading/hard/q4/RectangleOverload.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package methods.method_overloading.hard.q4;
+
+/*
+Write a Java Program that prompts the user to enter two integers and two doubles. Then write four overloaded methods
+called rectangleArea. The first takes both integer values for each side of the rectangle. The second takes the first
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/RectangleOverload.java:5]: We should probably add something about the goal of the question being to calculate the area of the rectangles based on the inputs (I feel like it is inferred, but not specifically stated.

File: src/methods/method_overloading/hard/q4/RectangleOverload.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+integer and the first double for each side of the rectangle. The third takes the second integer and uses it for all
+sides of the rectangle. The fourth takes the second double and uses it for all sides of the rectangle.
+
+Sample Output:
+  Enter two integers:
+  2 4
+  Enter two doubles:
+  5.3 3.3
+  The area of the rectangle is: 8
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/RectangleOverload.java:14]: We should also maybe change these to be like area of the first rectangle, second, etc. so the students have a better idea of what order they should print them in.

File: src/methods/method_overloading/hard/q4/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                Arguments.of(4, 76, 23.4246, 648372.1, 304, 93.6984, 5776, 4.2038638005841E11),
+                Arguments.of(32917, 4, 57.3, 1.321, 131668, 1886144.1, 16, 1.745041),
+                Arguments.of(-43, -23, 43, 763, 989, -1849.0, 529, 582169.0),
+                Arguments.of(4738, -32, -4372.432, 584.432391, -151616, -2.0716582816E7, 1024, 341561.21964997693)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("twoIntegersInputProvider")
+    void twoIntegerCorrect(int a, int b, int area) throws Throwable {
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/MainTest.java:98]: These test names are bad

File: src/methods/method_overloading/hard/q4/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        Object[][] arguments = {
+                {a, int.class},
+                {b, int.class}
+        };
+        MethodTest m = new MethodTest(RectangleOverload.class, "rectangleArea", arguments);
+        Object output = m.callMethod();
+        String errorMsg = "Your rectangleArea() method does not calculate the correct area for two integers.";
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/MainTest.java:105]: Should change the error messages to be "using two integers" instead of for (do this for all the messages)

File: src/methods/method_overloading/hard/q4/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        String errorMsg = "Your rectangleArea() method does not calculate the correct area for one double.";
+        CustomAssertions._assertEquals(area, output, 0.0001, errorMsg);
+    }
+
+    @ParameterizedTest
+    @MethodSource("mainMethodInputProvider")
+    void printsCorrectMessage(int a, int b, double x, double y, int twoInts, double intDouble, int oneInt, double oneDouble) {
+        runWithInput(a + " " + b + " " + x + " " + y);
+
+        assertEquals(twoInts, Integer.parseInt(getItemByName("int_int")), "Your program does not print the correct message for the first area.");
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/MainTest.java:151]: Should be "... correct output for the area of the first rectangle". Apply this to all the error messages.

File: src/methods/method_overloading/hard/q4/RectangleOverload.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
+package methods.method_overloading.hard.q4;
+
+/*
+Write a Java Program that prompts the user to enter two integers and two doubles. Then write four overloaded methods
+called rectangleArea. The first takes both integer values for each side of the rectangle. The second takes the first
+integer and the first double for each side of the rectangle. The third takes the second integer and uses it for all
+sides of the rectangle. The fourth takes the second double and uses it for all sides of the rectangle.
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/RectangleOverload.java:7]: Also, I think we should tell them to assume positive inputs.

File: src/methods/method_overloading/hard/q4/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                Arguments.of(2, 2.2, 4.4),
+                Arguments.of(-1, 1.12, -1.12),
+                Arguments.of(-1, -1.0, 1.0),
+                Arguments.of(112, 2222.123, 248877.776));
+    }
+
+    static Stream<Arguments> oneDoubleInputProvider() {
+        return Stream.of(
+                Arguments.of(1.2, 1.44),
+                Arguments.of(-1.0, 1.0),
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/MainTest.java:82]: Get rid of the negative inputs

File: src/methods/method_overloading/hard/q4/RectangleOverload.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
+package methods.method_overloading.hard.q4;
+
+/*
+Write a Java Program that prompts the user to enter two integers and two doubles which will be used to calculate
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/RectangleOverload.java:4]: We also need to specify the return type for the methods

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/methods/method_overloading/hard/q4/RectangleOverload.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+rectangle areas. This calculation will be performed by four overloaded methods called rectangleArea. The first takes
+both integer values for the sides of the rectangle and returns and integer. The second takes the first integer and the
```
Comment (Hedgemon4) [inline src/methods/method_overloading/hard/q4/RectangleOverload.java:6]:
```suggestion
both integer values for the sides of the rectangle and returns an integer. The second takes the first integer and the
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]
